### DISCUSSION TITLE
Systray implementation design

### DISCUSSION THREAD

**Discussion by proneon267** [beeware/toga#2108] [Ideas]
I will implement systray support. But, I have not forgotten about #1890, which I will be implementing later on. 

As the systray implementation is overlapping with notification implementation(especially on winforms), hence I think that a new class is needed. This will represent the app's icon in the System Tray(Windows) or status bar or notification area(for mobile). 

I am not sure about the naming of the class, as calling it systray icon would be confusing for other platforms. Hence, I have came up with `NotificationAreaIcon` or `StatusAreaIcon`.

The App class will have an instance of the `StatusAreaIcon` that will be automatically instantiated(with default settings) on the first access of `Toga.App.StatusAreaIcon`. 

The following will be the members and methods of `StatusAreaIcon`:
* `icon` - Gets and sets the icon of type `Toga.Image()`.
* `description` - Gets and sets the string representing the current status of the app or the progress of the work that the app is doing currently. Or more specifically the tooltip of the `StatusAreaIcon`.
* `show` - boolean to show/hide the `StatusAreaIcon`
* `add_to_menu("Name of the option", on_press=None)` - Adds a single option to the `StatusAreaIcon` menu with a handler.
* `add_to_menu(options=[("Name of the option1", on_press=None), ("Name of the option2", on_press=None),..])` - Adds multiple options to the `StatusAreaIcon` menu.
* `remove_from_menu("Name of the option")` - Remove an option from the menu.

What else do you think that needs to be in this API?
Related: #97 #1890 #907

**Comment by freakboy3742** [comment]
We're highly unlikely to add a "menu" API when we've managed to develop an entire App and Toolbar API that allows for app menus without once using the word "menu" in an API.

Another big gap in this API proposal is the "systray-only app". This proposal seems to be aimed at apps that have a main window *and* have a systray icon, but the much more pressing use case I can see is an app whose systray *is* the "main window".

**Reply by proneon267** [reply]
Well in that case, how does `add_option()` and `remove_option()` sound like?

As for "systray-only" app, I agree with you. I hadn't considered an app without any `MainWindow`. Is there any particular reason that the app can't just hide the `MainWindow` when it wants to be "systray-only"? 

Generally, most systray apps( like apps that operate entirely in systray like a "cpu status" app that's in systray) provide an option in their context menu to bring up their MainWindow. 

So, would showing and hiding the MainWindow would not be sufficient?

**Comment by freakboy3742** [comment]
> Well in that case, how does `add_option()` and `remove_option()` sound like?

The issue isn't the name. It's the entire pattern you're proposing for constructing a menu. 

3 questions for you that might be illuminating: 

1. How do you add a menu to a Toga app? 
2. How do you add a button to a Toga toolbar?
3. Why would the API for adding a menu to an app, or a button to a toolbar, be substantially different to the API for adding a menu to a systray?

> As for "systray-only" app, I agree with you. I hadn't considered an app without any `MainWindow`. Is there any particular reason that the app can't just hide the `MainWindow` when it wants to be "systray-only"?

Yes, there's a very good reason. A macOS app isn't required to have a Main Window. I believe the same is true of GTK as well.

In terms of Toga's API - in the case of a DocumentApp, there literally *isn't* a main window. There's a window for each *document*; and when you start the app, there isn't necessarily a document open.

> Generally, most systray apps( like apps that operate entirely in systray like a "cpu status" app that's in systray) provide an option in their context menu to bring up their MainWindow.

That's certainly a *common* pattern, but I wouldn't say it's universal. I've got several examples of apps in my system tray that don't have anything you could call their "main" window. The only element I'd say is universal is the *preferences* dialog - but that's not a main window.

**Reply by proneon267** [reply]
Just so we are clear, we are both talking about the context menu of the app when it is in systray, right?

Are you proposing to use similar API that's used for adding commands to the toolbar?Like the API should use `Toga.Command()` and the user can do `Toga.App.StatusAreaIcon.commands.add()`?

As for the "systray-only" aspect, I agree with you that the pattern of having a MainWindow isn't universal. Hence, I will do more research on this aspect and will propose an API design.

**Reply by freakboy3742** [reply]
> Just so we are clear, we are both talking about the context menu of the app when it is in systray, right?

Yes. Especially since the context menu of a "systray app" is the main UI of a systray app.
 
> Are you proposing to use similar API that's used for adding commands to the toolbar?Like the API should use `Toga.Command()` and the user can do `Toga.App.StatusAreaIcon.commands.add()`?

That's a lot closer to the style of API that I'd expect (i.e., creating commands and adding them to a CommandSet). However I'd suggest the API is going to be closer to `toga.App.systray.add()` - *if* a separate API is needed at all. 

To support the case of a "main window"-less app, we're likely going to need a standalone "toga.SystrayApp" - in which case, the commands on the App likely *are* the commands on the systray.

**Reply by proneon267** [reply]
I agree a standalone `toga.SystrayApp` is required. 

Also, what about the naming? What should it be called? `Systray`?

**Reply by freakboy3742** [reply]
"System tray" is about as close as we're going to get to a shared name for this kind of feature. I'd lean towards `system_tray`/`SystemTrayApp` rather than abbreviating to `systray`.

**Reply by proneon267** [reply]
> To support the case of a "main window"-less app, we're likely going to need a standalone "toga.SystrayApp" - in which case, the commands on the App likely are the commands on the systray.

That takes care of the API for adding options to the context menu, but what should be the behavior when the user wants the SystemTrayApp to behave like a Normal App and show a MainWindow?

There might be 3 situations:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|->**`[Window Present]`** & **`[SystemTray icon Absent]`**
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|->**`[Window Present]`** & **`[SystemTray icon Present]`**
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|->**`[Window Absent]`** & **`[SystemTray icon Present]`**

Does that mean separate instances of `Toga.App` and `Toga.SystemTrayApp` will be needed to run in parallel, when both the Window and SystemTray will be present?

**Reply by mhsmith** [reply]
How about "StatusIcon" as a neutral name? I feel like "systray" is a bit Windows-centric, and the "area" in "StatusAreaIcon" is redundant.

> To support the case of a "main window"-less app, we're likely going to need a standalone "toga.SystrayApp" - in which case, the commands on the App likely _are_ the commands on the systray.

Why would this need a separate class? On macOS, the commands on the app appear in the menu bar at the top left when the app is focused, while the commands on the status icon appear when you click on the icon at the top right when any app is focused. They're two separate things.

Also, we should probably design the API to give the option of creating multiple icons.

**Comment by freakboy3742** [comment]
> How about "StatusIcon" as a neutral name? I feel like "systray" is a bit Windows-centric, and the "area" in "StatusAreaIcon" is redundant.

I agree it's Windows specific; the only thing I'd say in the defence of that name is that the 3 or so times it's come up, the request has been for a "System tray icon". However, that may be because the people requesting it are Windows users.

I could live with StatusIcon though - a quick straw poll of some friends all hit on "Status bar" as the name of the place where these icons live; and Material Design guidelines apparently calls that area a Status Bar.

> > To support the case of a "main window"-less app, we're likely going to need a standalone "toga.SystrayApp" - in which case, the commands on the App likely _are_ the commands on the systray.
> 
> Why would this need a separate class? On macOS, the commands on the app appear in the menu bar at the top left when the app is focused, while the commands on the status icon appear when you click on the icon at the top right when any app is focused. They're two separate things.

Yes - but a status bar app doesn't need to have a main window. Dropbox and Docker are good examples here - in both cases, you can interact with the app without any window being open, anything appearing in the Dock, or the app menu changing. In the case of Docker, you only get a change in the App menu when you open the Docker Desktop Dashboard.

So - we need to be able to create an app that *doesn't* have a main window - and toga.App *requires* a main window. In the case of Windows, the MainWindow is actually critical to event processing, so we need an entirely different app startup mechanism.

> Also, we should probably design the API to give the option of creating multiple icons.

Can you give an example of a use case? I can't think of any apps that have multiple status bar icons.

**Reply by proneon267** [reply]
Re:
> To support the case of a "main window"-less app, we're likely going to need a standalone "toga.SystrayApp" - in which case, the commands on the App likely are the commands on the systray.

That takes care of the API for adding options to the context menu, but what should be the behavior when the user wants the SystemTrayApp to behave like a Normal App and show a MainWindow?

There might be 3 situations:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|->**`[Window Present]`** & **`[SystemTray icon Absent]`**
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|->**`[Window Present]`** & **`[SystemTray icon Present]`**
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|->**`[Window Absent]`** & **`[SystemTray icon Present]`**

Does that mean separate instances of `Toga.App` and `Toga.SystemTrayApp` will be needed to run in parallel, when both the Window and SystemTray will be present?

**Reply by freakboy3742** [reply]
No - whatever API is used to define the icon would be on both classes. If you want an "normal" app with a MainWindow, subclass toga.App; if you want a `Systray`/`StatusIcon`-only app, subclass `toga.StatusIconApp`. Having common functionality across multiple classes is literally why class inheritance exists as a concept.

**Reply by proneon267** [reply]
>No - whatever API is used to define the icon would be on both classes. If you want an "normal" app with a MainWindow, subclass toga.App; if you want a Systray/StatusIcon-only app, subclass toga.StatusIconApp. Having common functionality across multiple classes is literally why class inheritance exists as a concept.

That makes sense. What type of common API do you have in mind for adding options to context menu, that would be intuitive both for anyone who either uses `Toga.App` or a `Toga.StatusIconApp`? 

Maybe we can add a `Toga.App.StatusIcon.add()` on the `Toga.App` class and on the `Toga.StatusIconApp` class, we can use the added commands as options in the context menu.

>So - we need to be able to create an app that doesn't have a main window - and toga.App requires a main window. In the case of Windows, the MainWindow is actually critical to event processing, so we need an entirely different app startup mechanism.

Its not entirely true that there is a hard requirement for a `MainWindow`. On WinForms, if you comment out:
https://github.com/beeware/toga/blob/037084b1126bacc3128aa38d390b349b0f8a9718/winforms/src/toga_winforms/app.py#L137
https://github.com/beeware/toga/blob/037084b1126bacc3128aa38d390b349b0f8a9718/winforms/src/toga_winforms/app.py#L138
and prevent the instance checking of `self.main_window` by commenting out:
https://github.com/beeware/toga/blob/037084b1126bacc3128aa38d390b349b0f8a9718/core/src/toga/app.py#L557

 Now, even if you don't define `self.main_window` in the App's `startup()` method then also the App's event loop starts.
 
Infact, Now assigning a `Toga.Window()` instead of a `Toga.MainWindow()` to `self.main_window` also works and shows the window correctly. 

Also, why is `create_menus()` on the `Toga.App` class instead of being in the `MainWindow()` class?  The toolbar menu is a Window specific feature, hence any related APIs should be on the Window class.

This reminds me, maybe we should discuss more about an app class that doesn't have any windows, but does have an event loop and other Application specific APIs. This would then be inherited and extended by other types of App Classes that implement Window functionality.

**Reply by freakboy3742** [reply]
> > No - whatever API is used to define the icon would be on both classes. If you want an "normal" app with a MainWindow, subclass toga.App; if you want a Systray/StatusIcon-only app, subclass toga.StatusIconApp. Having common functionality across multiple classes is literally why class inheritance exists as a concept.
> 
> That makes sense. What type of common API do you have in mind for adding options to context menu, that would be intuitive both for anyone who either uses `Toga.App` or a `Toga.StatusIconApp`?
> 
> Maybe we can add a `Toga.App.StatusIcon.add()` on the `Toga.App` class and on the `Toga.StatusIconApp` class, we can use the added commands as options in the context menu.

Well - for starters, Python uses CamelCase to describe classes, not properties - so at the very least, it should be `toga.App.status_icon.add()`. 

Beyond that - you're the person proposing to work on this feature. It's up to you to propose something. And that proposal is more than "this is what the API is called" - it's how that will be interpreted on various platforms, and demonstrating that the API will be sufficient to reach the use cases for the feature.

> > So - we need to be able to create an app that doesn't have a main window - and toga.App requires a main window. In the case of Windows, the MainWindow is actually critical to event processing, so we need an entirely different app startup mechanism.
> 
> Its not entirely true that there is a hard requirement for a `MainWindow`. On WinForms, if you comment out:
...
> Now, even if you don't define `self.main_window` in the App's `startup()` method then also the App's event loop starts.

Are you sure about that?

https://github.com/beeware/toga/blob/main/winforms/src/toga_winforms/libs/proactor.py#L77

Add a background task to your app, and comment out your main window as you describe. The background task won't run.
 
> Also, why is `create_menus()` on the `Toga.App` class instead of being in the `MainWindow()` class? The toolbar menu is a Window specific feature, hence any related APIs should be on the Window class.

As with any question about this sort of design, the easiest way to reverse-engineer the answer is to look for platform differences. In this case - macOS is the platform that points out the reason. On macOS, menus are a property of an *app*, not a window.

However, although it might not be obvious, this is fundamentally true of other platforms too. Yes, on Windows and GTK, the menu bar is physically contained on a window - but it is only *ever* on the *main* window. You don't put menus on non-main windows. An App has menus; they might manifest on the main window, depending on the platform.

You do, however put *toolbars* on non-main windows, so the toolbar API is window-specific.

> This reminds me, maybe we should discuss more about an app class that doesn't have any windows, but does have an event loop and other Application specific APIs. This would then be inherited and extended by other types of App Classes that implement Window functionality.

I'm not sure what there is to discuss. Yes, if there is common functionality, it should be factored into a base class. Again, this is why inheritance exists. 

---

As a meta comment: I don't want to discourage your sprit of contribution, but I would like to suggest that you might want to re-evaluate the types of projects you're taking on. 

You currently have 3 major design projects in flight - this discussion about System trays, plus Screens, and Window focus events. You've started and abandoned a couple of others. In each case, you're tackling a large greenfields API design, and to be brutally frank, the designs you're proposing don't reflect familiarity with existing design patterns, platform considerations, or naming conventions in the Toga codebase. We're doing what we can to help you get these designs to a final state, but this is a time consuming process.

You've then provided skeleton implementations for review that, in several cases, have failed basic "does this start?" testing. 

You *have* submitted a number of PRs that we have merged. From what I can tell, you have access to a Windows machine, and maybe a GTK machine as well - and the PRs that we *have* merged have all been in the territory of smaller, platform-specific bug fixes on platforms to which you have access. 

Can I suggest that you focus your efforts on incremental improvement to existing APIs, rather than developing large scale new APIs - at least until you're more familiar with Toga. We have no shortage of bugs logged in the Windows and GTK backends, and the audit process has revealed plenty of features (especially on the Winforms backend) that are lacking implementations, or have partial implementations. We're planning to turn these into individual tickets once the audit is complete, but in the meantime, you can search for skipped tests in the test backends. These are all areas where we have established APIs, and what we're lacking is an implementation. The work is well scoped, the design is known, success criteria are already established - we just need someone to work out the implementation details for a single platform.

At the very least, could I ask you to work on one project at a time? The Screens API (#1930) is the closest to completion, but it still has a *lot* of work to be done; yes, that testing is dependent on on #2075 - but you *can* start that work now by building a git branch that merges #2075 and #1930. Once #2075 lands, you can rebase your testing branch.

Again, I don't want to discourage you from contributing to Toga - I want to find a way for you to contribute that is better aligned with your skills and resources.

**Reply by proneon267** [reply]
I completely understand your point for smaller, more incremental contributions. I agree and will focus on making these contributions to the project moving forward. I will complete the existing PRs that I have opened and will not start any new designs. Thanks for the guidance! :)

**Reply by mhsmith** [reply]
> So - we need to be able to create an app that _doesn't_ have a main window - and toga.App _requires_ a main window. In the case of Windows, the MainWindow is actually critical to event processing, so we need an entirely different app startup mechanism.

OK, I admit I'm not too familiar with the WIndows event loop, but if possible I think we should try to fix those limitations rather than baking them into the API by adding separate classes.

> > Also, we should probably design the API to give the option of creating multiple icons.
> 
> Can you give an example of a use case? I can't think of any apps that have multiple status bar icons.

I've seen system monitor apps for both Windows and macOS which can show one icon for CPU load, one for network activity, etc.

**Reply by proneon267** [reply]
> OK, I admit I'm not too familiar with the WIndows event loop, but if possible I think we should try to fix those limitations rather than baking them into the API by adding separate classes.

@mhsmith This is fixed by #2112. If possible, can you try it out and tell if the closing behavior feels right to you? There is a need for a design decision on "when should the app close?", now that MainWindow is not critical for event processing.

**Reply by freakboy3742** [reply]
> > So - we need to be able to create an app that _doesn't_ have a main window - and toga.App _requires_ a main window. In the case of Windows, the MainWindow is actually critical to event processing, so we need an entirely different app startup mechanism.
> 
> OK, I admit I'm not too familiar with the WIndows event loop, but if possible I think we should try to fix those limitations rather than baking them into the API by adding separate classes.

I pointed that out more as a notable implementation detail than as a supporting reason for the extra base class. 

From an API perspective, what I'm highlighting is that, independent of any platform requirements, `toga.App` currently has a hard requirement that an app has a main window, and closing that main window is a condition for exiting the app. A "pure" status icon app *won't* have a main window, but an app with a main window *might* have one or more status icons.

If we can find a way to express this with a single App class, then I'm all in favor - but it isn't immediately obvious to me how we can do this without impacting on the usability of App for the "normal" use case.

> > > Also, we should probably design the API to give the option of creating multiple icons.
> > 
> > Can you give an example of a use case? I can't think of any apps that have multiple status bar icons.
> 
> I've seen system monitor apps for both Windows and macOS which can show one icon for CPU load, one for network activity, etc.

Fair enough. It's a niche requirement, but unless it overly complicates the API design, I can't see any reason we shouldn't support it.

From a quick look into the Cocoa API, the API terminology actually aligns nicely with this nicely - there's a "status bar" that has 0+ "status items", each of which is a button that can optionally have a menu.

**Reply by mhsmith** [reply]
> @mhsmith This is fixed by #2112. If possible, can you try it out and tell if the closing behavior feels right to you? 

Although I have more recent experience as a Windows *user*, @freakboy3742 still knows more than me about the low levels of the Windows Toga implementation, so I'll leave that PR to him.

**Reply by proneon267** [reply]
I agree :)

**Reply by proneon267** [reply]
>If we can find a way to express this with a single App class, then I'm all in favor - but it isn't immediately obvious to me how we can do this without impacting on the usability of App for the "normal" use case.

I will research more and propose an API design that would comply with this idea.

**Reply by proneon267** [reply]
**Proposal for the API design:**
We should treat StatusIcons like Windows. This would allow us to not introduce a new App class and to have multiple StatusIcons. The app class would have a StatusIcon registry(`toga.App.status_icons`) similar to the Window registry(`toga.App.windows`). The options would be added by the existing `toga.Command` Resource.

**To instantiate the class:**
```
self.status_icon = toga.StatusIcon(icon: toga.Icon=None, tooltip: str=None)
```

**To add an option to the context menu of the StatusIcon:**
```
self.status_icon.options.add(*commands: toga.Command)
```

**To remove an option from the context menu of the StatusIcon:**
```
self.status_icon.options.remove(*commands: toga.Command)
```

**To associate the StatusIcon with the app:**
```
toga.App.status_icons += self.first_status_icon
```

```
toga.App.status_icons.add(*statusicons: toga.StatusIcon)
```

**To show the StatusIcon:**
```
self.status_icon.show()
```

**To hide the StatusIcon:**
```
self.status_icon.hide()
```

An example usage would be:

**For apps having only StatusIcons:**
```
def startup(self):

        self.first_status_icon = toga.StatusIcon(tooltip="First StatusIcon")
        self.status_icons += self.first_status_icon
        self.first_status_icon.options.add(cmd1, cmd2)
        self.first_status_icon.show()
        
        self.second_status_icon = toga.StatusIcon(tooltip="Second StatusIcon")
        self.status_icons += self.second_status_icon
        self.second_status_icon.options.add(cmd3, cmd4)
        self.second_status_icon.show()
```

Note here that there is no `main_window`. This is for creating "StatusIcon-only" apps. 

**For apps having both StatusIcons and windows:**
```
 def startup(self):
        self.main_window = toga.MainWindow(title="Main Window")
        self.main_window.show()

        self.second_window = toga.Window(title="Second Window")
        self.windows += self.second_window
        self.second_window.show()

        self.first_status_icon = toga.StatusIcon(tooltip="First StatusIcon")
        self.status_icons += self.first_status_icon
        self.first_status_icon.options.add(cmd1, cmd2)
        self.first_status_icon.show()
        
        self.second_status_icon = toga.StatusIcon(tooltip="Second StatusIcon")
        self.status_icons += self.second_status_icon
        self.second_status_icon.options.add(cmd1, cmd2)
        self.second_status_icon.show()
```

**For apps having only windows:**
```
 def startup(self):
        self.main_window = toga.MainWindow(title="Main Window")
        self.main_window.show()

        self.second_window = toga.Window(title="Second Window")
        self.windows += self.second_window
        self.second_window.show()
```


Any changes and suggestions are welcome :)

**Reply by freakboy3742** [reply]
Once again, I'm again asking you to restrict yourself to one major new piece of work at a time. You have three major pieces in work in flight at present; and despite your statement that you understood and agreed that finishing the existing work was necessary before starting new things, you're continuing to develop designs here. Please focus on *one* of the major projects - preferably one that has already gotten through the design process - before continuing the design process on others.

For the sake of posterity - I have 2 critiques of this proposed design:
1. It's not clear to me why toga.StatusIcon is needed. We created an entire toolbar API without adding toga.ToobarButton or similar - why is toga.StatusIcon not just a toga.Group, and app.status_icons a normal toga.CommandSet?
2. You say we should "treat StatusIcon like a window" - but I can't see anything in this proposal where you... treat it like a window. It's a completely separate attribute. It's not added to `app.windows`. It doesn't take over the role of `app.main_window`. In what way is it being "treated like a window"? As I've indicated elsewhere, my fundamental concern here is that this API proposal requires changing the contract of toga.App so that you can no longer assume that `main_window` exists. 
 
The second point *is* something we can change if we choose to do so - but we need to be careful that we've actually thought through the consequences. One big reason for having a different subclass: on macOS at least, the icon in the Dock is based on the *app*, not the windows of the app. In a "toolbar only" app, you don't want a Dock Icon, which requires changing the entire App execution mode - and the App is the *first* thing that is constructed, long before you've executed any code that creates a `main_window`.

**Reply by proneon267** [reply]
>Once again, I'm again asking you to restrict yourself to one major new piece of work at a time. You have three major pieces in work in flight at present; and despite your statement that you understood and agreed that finishing the existing work was necessary before starting new things, you're continuing to develop designs here. Please focus on one of the major projects - preferably one that has already gotten through the design process - before continuing the design process on others.

Apologies, this will be my last reply on this thread, until I have completed all the existing opened PRs.

"treat StatusIcon like a window" - I did not meant that literally. `toga.StatusIcon()` will be a separate class. Its job will be to essentially take the role of a `toga.Window()` class. Like adding commands to the window, adds the commands to the toolbar. Similarly, adding commands to the `toga.StatusIcon()` would add the commands to the context menu of the StatusIcon. The construction of the StatusIcon's context menu would be similar to the way the commands are added to the app menu, meaning the existing patterns of `toga.Group`, `toga.Command`, etc will be used. Hence, creation of any new pattern of adding options would be avoided.

>on macOS at least, the icon in the Dock is based on the app, not the windows of the app. In a "toolbar only" app, you don't want a Dock Icon, which requires changing the entire App execution mode - and the App is the first thing that is constructed, long before you've executed any code that creates a main_window.

I will research more and find a solution, but this feels like something we can solve by passing a flag parameter to the App class during its initialization to indicate, if the dock icon should be visible or not. In fact, this case is not specific to macOS only, other platforms would also need to not show app icon for a "toolbar only" app. 

But, I will not continue to work on this, until I have completed the existing opened PRs.
